src/lexer.rs

```diff
@@ -0,0 +1,432 @@
+use crate::{
+    asciifile::{Position, PositionedChar, PositionedChars},
+    strtab::*,
+};
+
+use std::{convert::TryFrom, result::Result};
+
+macro_rules! match_op {
+    ($input: expr, $len: expr, $right: expr) => {{
+        // Unwraps are safe, because this is only called after token of length $len,
```

Comment:
I guess that would be the "Ansatz 2" suggested on the slides. I don't think it would be that hard to implement this inside the `PositionedChars` iterator (i already have a buffer for the peeked values), just turn this around a bit. 

This would be more problematic in the lexer itself, I think: If e.g. the `.peek()` in line 253 (`lex_token`) would already consume the first digit of the integer, `lex_integer_literal` would miss it's first digit. There are probably equally many arguments in the code for both sides.

But turning it around would get rid of the (safe) unwraps... Hm.

Reply:
Also returning the cursor position in `peek()` would make the lexer code a bit more obfuscated, avoiding this was the originial reason I added a custom `peek` function. At the very least, almost every peek would have to be trailed with `.map(|v| v.2)`.

Reply:
orthogonal, but I'm afraid my upcoming commit of the following diff will close this comment:
```
diff --git a/src/lexer.rs b/src/lexer.rs
index 75c7164..0d591c9 100644
--- a/src/lexer.rs
+++ b/src/lexer.rs
@@ -17,14 +17,10 @@ macro_rules! match_op {
     ($input:expr, $len:expr, $right:expr) => {{
         // Unwraps are safe, because this is only called after token of length $len,
         // is already matched and thus contained in $input
-        let PositionedChar(begin, _) = $input.next().unwrap();
-        let mut end = begin;
-        #[allow(clippy::reverse_range_loop)] // Macro might be called with with $len=1
-        for _ in 1..$len {
-            let PositionedChar(pos, _) = $input.next().unwrap();
-            end = pos;
-        }
-
+        assert!($len >= 1);
+        let mut it = $input.by_ref().take($len);
+        let PositionedChar(begin, _) = it.next().unwrap();
+        let end = it.last().map(|pc| pc.0).unwrap_or(begin);
         Some(Ok(Token::new(begin, end, TokenKind::Operator($right))))
     }};
 }
```

Reply:
ok, github was clever

Reply:
As far as I see it, this conversation is the only one left. May the stakeholders decide...